<div >
<img src = "figs/ans_banner_1920x200.png" />
</div>

# Caso-taller:  Recomendando el Blog de  Hernán Casciari 


[Hernán Casciari](https://hernancasciari.com/#bio), es un escritor argentino, que escribe blog posts con cuentos e historias  relacionadas con el futbol, su vida, infancia, y relaciones familiares con toques de ficción. Este [blog](https://hernancasciari.com/blog/) es  tan interesantes que en 2005 fue premiado como “El mejor blog del mundo” por Deutsche Welle de Alemania. 

El objetivo de este caso-taller es construir un sistema de recomendación basado en los contenidos de los posts utilizando similitud de las palabras usadas o temas de los cuentos.

## Instrucciones generales

1. Para desarrollar el *cuaderno*, primero debe descargarlo.

2. Para responder cada inciso deberá utilizar el espacio debidamente especificado.

3. La actividad será calificada sólo si sube el *cuaderno* de jupyter notebook con extensión `.ipynb` en la actividad designada como "entrega calificada por el personal".

4. El archivo entregado debe poder ser ejecutado localmente por el tutor. Sea cuidadoso con la especificación de la ubicación de los archivos de soporte, guarde la carpeta de datos en el mismo `path` de su cuaderno, por ejemplo: `data`.

## Desarrollo


### 1. Carga de datos 

En la carpeta `data` se encuentran el archivo `blog_casciari.csv` con el título, la fecha de publicación, y el contenido de los cuentos publicados en el blog  de sr. Casciari. Cargue estos datos en su *cuaderno* y reporte brevemente el contenido de la base.
   

In [16]:
# Utilice este espacio para escribir el código.
import pandas as pd
import numpy as np

# Convertir el csv a un dataframe usando pandas
casciari = pd.read_csv("data/blog_casciari.csv", sep = ",")

# Ilustrar la esturctura del dataframe
print(casciari.head())
# Exploramos la estructura del dataframe
print("La estructura del dataframe es de dimensión", casciari.shape)

# Vemos el cuento ejemplo - 160 La venganza del metegol
#example=casciari['cuento'][160]
print(casciari['titulo'][160])
print(casciari['cuento'][160])


                       titulo    fecha  \
0            El rincón blanco  1/11/08   
1  Mínimos avances en la cama  1/24/08   
2                  Don Marcos  2/19/08   
3              Los dos rulfos  3/26/08   
4   La noticia no es el perro  4/15/08   

                                              cuento  
0  De pronto yo estaba en el hogar donde pasé la ...  
1  Menos la cama, todo ha mejorado en este mundo....  
2  Dos veces, y no una, mi abuelo materno me ayud...  
3  A su regreso de México, mi amigo Comequechu no...  
4  De repente, un video de You Tube recibe un mil...  
La estructura del dataframe es de dimensión (520, 3)
La venganza del metegol
El mes pasado me invitaron a presentar un libro en Buenos Aires. Y como era un libro sobre fútbol, al final de la charla el director de la editorial nos invitó a jugar un partido de metegol (ese invento español al que sus creadores llaman, erróneamente, futbolín). Hacía años que no jugaba al metegol, pero por suerte me tocó de compañero u

#### Procedimiento:

1. Importamos las librerias a utilizar, en este caso son pandas y numpy
2. Convertimos el archivo csv a un dataframe utilizando pandas
3. Exploramos la estructura del dataframe
4. Usamos la función shape para visualizar las dimensiones del dataframe

#### Análisis y conclusiones: 

Podemos ver que el Blog de Hernán Casciri contiene 520 entradas y el archivo nos provee la información de título y fecha y cuento. 

### 2. Homogenización de textos

Para cumplir con el objetivo de generar recomendaciones en esta sección debe preparar los posts para poder ser utilizados en su sistema de recomendación. Para ello, "limpie" y "tokenize" cada uno de los cuentos, describiendo detalladamente los pasos que realizo y si transformó o eliminó ciertas palabras. Para asistirlo en la tarea he creado listas de *stopwords* que están disponibles en la carpeta `data`. En su procedimiento ilustre la limpieza con el cuento 'La venganza del metegol'. (En su limpieza recuerde que el objetivo es generar recomendaciones a partir de la similitud de las palabras o temas de los cuentos)

In [12]:
### Limpieza y tokenización siguiendo el ejemplo del LAB Intro al PLN

# PENDIENTES:
# 1) FALTA USAR LA LISTA DE STOPWORDS PROVISTA PARA EL TALLER
# 2) FALTA EN CADA PASO MOSTRAR LA TRNASFORMACIÓN DEL BLOG "LA VENGANZA DEL METEGOL"


# Cargamos las librerías a utilizar
import re
import unidecode
import spacy
from wordcloud import WordCloud
import matplotlib.pyplot as plt


#Escogemos el cuento la venganza del metegol como ejemplo
example=casciari['cuento'][160]
example


# Cargar el modelo para el idioma deseado (por ejemplo, español)
nlp = spacy.load("es_core_news_sm")

## Retiramos caracteres no deseados

# Quitamos tildes
documento = unidecode.unidecode(casciari)
print(documento[5,:,:]) 

# Retiramos caracteres no deseados
documento = re.sub('[^A-Za-z0-9 ]+', ' ', documento)

# Eliminamos espacios extras
documento = re.sub('\s+', ' ', documento)

# Minúsculas
documento = documento.lower()

## Tokenización
doc = nlp(documento)


# Filtrar las stopwords
doc_sin_stopwords = [token.text for token in doc if not token.is_stop]

# Convertir la lista de palabras sin stopwords nuevamente a texto
doc_sin_stopwords = ' '.join(doc_sin_stopwords)

# Obtener los lemas de cada palabra
lemmas =[token.lemma_ for token in nlp(doc_sin_stopwords)]


# Convertir la lista de lemmas nuevamente a texto
doc_lemmas = " ".join(lemmas)

# Creamos el objeto WordCloud con el documento de lemas
wordcloud = WordCloud(width = 1600, height = 800, 
    background_color = "white").generate(doc_lemmas)


# Mostramos la nube de palabras usando Matplotlib
plt.figure(figsize = (20, 10))
plt.imshow(wordcloud, interpolation = 'bilinear')
plt.axis("off") # elimina los ejes y marcas para tener una visualización más limpia.
plt.show()



ModuleNotFoundError: No module named 'spacy'

#### Procedimiento:

1.

#### Análisis y conclusiones: 

1. 

### 3. Generando Recomendaciones

En esta sección nos interesa generar recomendaciones de cuentos en el blog a un usuario que leyó 'La venganza del metegol'. Para ello vamos a utilizar distintas estrategias.

#### 3.1. Recomendaciones basadas en contenidos

##### 3.1.1. Genere 5 recomendaciones de más recomendada (1) a menos recomendada (5) para el cuento 'La venganza del metegol' usando en la distancia de coseno donde el texto este vectorizado por `CountVectorizer`. Explique el procedimiento que realizó y como ordenó las recomendaciones.

In [ ]:
# Utilice este espacio para escribir el código.

(Utilice este espacio para describir el procedimiento, análisis, y conclusiones)

##### 3.1.2. Genere 5 recomendaciones de más recomendada (1) a menos recomendada (5) para  el cuento 'La venganza del metegol' usando nuevamente la distancia de coseno, pero ahora vectorice el texto usando `TF-IDFVectorizer`. Explique el procedimiento que realizó y como ordenó las recomendaciones. Compare con los resultados del punto anterior y explique sus similitudes y/o diferencias.

In [ ]:
# Utilice este espacio para escribir el código.

(Utilice este espacio para describir el procedimiento, análisis, y conclusiones)

##### 3.1.3. Genere 5 recomendaciones de más recomendada (1) a menos recomendada (5) para el cuento 'La venganza del metegol' usando el texto vectorizado por `TF-IDFVectorizer` y la correlación como medida de similitud. Explique el procedimiento que realizó y como ordenó las recomendaciones. Compare con los resultados de los puntos anteriores y explique sus similitudes y/o diferencias.

In [ ]:
# Utilice este espacio para escribir el código.

(Utilice este espacio para describir el procedimiento, análisis, y conclusiones)

##### 3.2. Recomendaciones basadas en temas

Usando modelado de temas con LDA, encuentre los temas subyacentes en el blog. Explique como eligió el numero óptimo de temas. Utilizando el tema asignado al cuento 'La venganza del metegol' y la probabilidad de pertenecer a este tema genere 5 recomendaciones de más recomendada (1) a menos recomendada (5) para este cuento. Explique el procedimiento que realizó. Compare con los resultados encontrados anteriormente y explique sus similitudes y/o diferencias. (Esto puede tomar mucho tiempo y requerir mucha capacidad computacional, puede aprovechar los recursos de [Google Colab](https://colab.research.google.com/))


In [ ]:
# Utilice este espacio para escribir el código.

(Utilice este espacio para describir el procedimiento, análisis, y conclusiones)

### 4 Recomendaciones generales

De acuerdo con los resultados encontrados, en su opinión ¿qué procedimiento generó las mejores recomendaciones para la entrada elegida? ¿Cómo implementaría una evaluación objetiva de estas recomendaciones? Justifique su respuesta.

(Utilice este espacio para describir su procedimiento)